Added Cloud Provider & Terraform Changes for Custom endpoint support

satya-umd opened the pull request:
Added changes to consume Install config and create AWS cloud provider config.
Also, added terraform changes to enable using custom endpoints along with the
necessary tests.

Story - https://jira.coreos.com/browse/CORS-1113

satya-umd commented:
/assign @patrickdillon

abhinavdahiya commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+// CloudConfig is the aws cloud provider config.
+type CloudConfig struct {
```
why is this public?

satya-umd replied:
Changed it the private.

abhinavdahiya commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	index := 1
+	for _, t := range params.CustomRegionOverride {
```
why can't you use the index from this range?

satya-umd replied:
Reworked the code and removed use of index.

abhinavdahiya commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+
+	index := 1
+	for _, t := range params.CustomRegionOverride {
+		s, err := file.NewSection(fmt.Sprintf("ServiceOverride \"%d\"", index))
```
is `%q` the better option?

satya-umd replied:
Updated code to use %q

abhinavdahiya commented on pkg/types/aws/customendpoint.go:
```diff
@@ -0,0 +1,11 @@
+// override existing defaults of AWS Services.
+// Currently Supports - EC2, IAM, ELB, S3 and Route53. 
+type CustomEndpoint struct{
+	Service       string `json:"service"`
```
would like to see field level comments

patrickdillon replied:
alternatively, I think you could consider not creating these types at all and just storing the endpoints as key value pairs in a map. I think this struct looks really nice and clean, but ultimately terraform wants the map. @abhinavdahiya what do you think? 

Also I think we might be missing some files. I was trying to explore this idea further but I don't see installconfig.go so I couldn't see how this struct is created?

patrickdillon replied:
My suggestion above would be a significant change... I think in order for the yaml to be unmarshalled to a map it would require the format of the install config to change (not sure if it is possible, but I assume it is).

satya-umd replied:
Added additional field level comments.

abhinavdahiya commented:
* This is missing the destroyer code to use the custom endpoints.
* all the other code for example, fetching availability zones, public zone ID etc.

satya-umd commented:
/cc cuppett

cuppett commented:
@jottofar may want to take a look as well. This is great to have land in the installer!

patrickdillon commented on data/data/aws/main.tf:
```diff
@@ -9,6 +9,14 @@ locals {
+    ec2 = lookup(var.custom_endpoints, "ec2", null)
+    elb =  lookup(var.custom_endpoints, "elb", null)
+    iam =  lookup(var.custom_endpoints, "iam", null)
+    route53 =  lookup(var.custom_endpoints, "route53", null)
```
terraform linting issue: these equal signs and values should be lined up.

satya-umd replied:
Updated terraform linting.

patrickdillon commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -134,6 +134,7 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		if err != nil {
 			return err
 		}
+		customEndpointsMap := installConfig.Config.Platform.AWS.FetchCustomEndpointsMap()
```
nit: I would not include the type in the name ([Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation))
i.e. something like
`customEndpointsMap` -> `customEndpoints`
`FetchCustomEndpointsMap` -> `FetchCustomEndpoints`

satya-umd replied:
Changed the variable names to remove type in the name.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+// global struct of CloudConfig which is currently not inialized.
+type global struct {
```
Did we determine if this struct is needed or not? What is the actual output in the cloud provider config considering that we're not populating any of the values?

satya-umd replied:
Update the Global struct and removed unused attributes.

satya-umd replied:
Removed all the unused attributes in the global struct.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+// CloudProviderConfig builds the cloud provider config and reflects to an ini file.
+func CloudProviderConfig(params *awstypes.Platform) (string, error) {
```
can we come up with a more descriptive name than `params`? You might also consider passing the overrides and region as separate params

satya-umd replied:
Renamed the variable to 'awsPlatform' instead.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	index := 1
+	for _, t := range params.CustomRegionOverride {
```
what is `t` short for? is there a better variable name?

satya-umd replied:
Renamed the variable to 'endpoint'.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+				Region:  params.Region,
+				URL:     t.URL,
+			}); err != nil {
+			return "", errors.Wrapf(err, "failed to reflect from  ServiceOverride")
```
extra space between `from` and `ServiceOverride`

patrickdillon commented on pkg/tfvars/aws/aws.go:
```diff
@@ -24,7 +25,7 @@ type config struct {
 
 // TFVars generates AWS-specific Terraform variables launching the cluster.
-func TFVars(masterConfigs []*v1beta1.AWSMachineProviderConfig, workerConfigs []*v1beta1.AWSMachineProviderConfig) ([]byte, error) {
+func TFVars(masterConfigs []*v1beta1.AWSMachineProviderConfig, workerConfigs []*v1beta1.AWSMachineProviderConfig, customEndpointsMap map[string]string) ([]byte, error) {
```
same as https://github.com/openshift/installer/pull/2073/files#r306984378

satya-umd replied:
Removed the type from the variable name.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+	Service       string `ini:"Service"`
+	Region        string `ini:"Region"`
+	URL           string `ini:"URL"`
+	SigningRegion string `ini:"SigningRegion,omitempty"`
```
Similar to https://github.com/openshift/installer/pull/2073/files#r306990159, we should remove the fields beginning with Signing... we aren't using them.

satya-umd replied:
Removed unused fields from the struct.

patrickdillon commented on pkg/types/aws/validation/customendpoint.go:
```diff
@@ -0,0 +1,50 @@
+			i++
+		}
+		sort.Strings(validValues)
+		return validValues
```
I'm not sure I get this. I could be missing something but isn't this just iterating through the map above and returning a []string of all the keys?

satya-umd replied:
List of services is returned in the 'value not supported' validation error condition, above implementation is similar to this -
https://github.com/openshift/installer/blob/7a05abd21e86ab75c5fc1b1ae042cc9a2940963b/pkg/types/aws/validation/platform.go#L40-L49

patrickdillon commented on pkg/types/aws/platform.go:
```diff
@@ -19,4 +19,20 @@ type Platform struct {
+	// CustomRegionOverride list contains custom endpoints which will overridde default
+	// service endpoint of AWS Services. 
+	// +optional
+	CustomRegionOverride []CustomEndpoint  `json:"customRegionOverride,omitempty"` 
```
I'm not sure how I feel about the name `CustomRegionOverride`. I think we borrowed that name from the k8s cloudproviderconfig, and that's fine. I think it is especially clunky in the install config. personally I like just endpoint but also perhaps customEndpoint.

patrickdillon commented on pkg/asset/manifests/aws/cloudproviderconfig.go:
```diff
@@ -0,0 +1,75 @@
+		if err := s.ReflectFrom(
+			&serviceOverride{
+				Service: t.Service,
+				Region:  params.Region,
```
endpoints can be in different regions than the cluster. it might be worth considering defaulting to the cluster region so that users don't have to fill it out in the install config, but they should be able to override this.

satya-umd replied:
Already discussed in design document, will not be updating this.

satya-umd commented:
> * This is missing the destroyer code to use the custom endpoints.
> * all the other code for example, fetching availability zones, public zone ID etc.

Addressed these changes and added code to handle for the above cases.

patrickdillon commented on pkg/destroy/aws/aws.go:
```diff
@@ -1098,19 +1122,22 @@ func deleteElasticLoadBalancing(session *session.Session, arn arn.ARN, logger lo
 			return errors.Errorf("unrecognized elastic load balancing resource subtype %s", subtype)
 		}
 	case "targetgroup":
+		//satya
```
can you delete or elaborate on this

patrickdillon commented on pkg/destroy/aws/aws.go:
```diff
@@ -1098,19 +1122,22 @@ func deleteElasticLoadBalancing(session *session.Session, arn arn.ARN, logger lo
 		id = segments[1]
 		switch subtype {
 		case "net":
+			//satya
```
can you delete or elaborate on this

patrickdillon commented on pkg/types/aws/platform.go:
```diff
@@ -19,4 +19,21 @@ type Platform struct {
+	// CustomRegionOverride list contains custom endpoints which will overridde default
+	// service endpoint of AWS Services.
+	// +optional
+	CustomRegionOverride []CustomEndpoint `json:"customRegionOverride,omitempty"`
```
if we have decided we're not going to allow users to specify a region for their custom endpoint, then I still feel we could greatly simplify this code by unmarshaling directly into a map[string]string`. 

Terraform only accepts the map[string]string. If further fields were needed for the cloud provider config then it could be hand edited in the manifest.

patrickdillon commented:
@satya-umd Looks like the tests are passing (scaleup is non-blocking)! Can you squash the review change commits? I would imagine you could squash this down to two commits: the original and adding destroy commands.

patrickdillon commented:
/retest
